src/s4.cpp

```diff
@@ -0,0 +1,636 @@
+  {
+    HTTPCode rc = remote_s4->handle_put(id, aor, trail);
+
+    if (rc == HTTP_PRECONDITION_FAILED)
+    {
+      // We've tried to do a PUT to a remote site that already has data. We need
+      // to send a PATCH instead.
+      TRC_DEBUG("Need to convert PUT to PATCH for %s on %s",
+                id.c_str(), _id.c_str());
+
+      PatchObject po;
+      convert_aor_to_patch(aor, po);
+
+      AoR* remote_aor = NULL;
+      remote_s4->handle_patch(id, po, &remote_aor, trail);
+      delete remote_aor; remote_aor = NULL;
+    }
+  }
+}
+
+// Replicate the PATCH to each remote site. We don't care about the return code
+// from the remote sites unless it's PRECONDITION FAILED, in which case we want
+// to send a PUT instead (to reinstantiate the subscriber).
+void S4::replicate_patch_cross_site(const std::string& id,
+                                    const PatchObject& po,
+                                    const AoR& aor,
+                                    SAS::TrailId trail)
+{
+  for (S4* remote_s4 : _remote_s4s)
+  {
+    AoR* remote_aor = NULL;
+    HTTPCode rc = remote_s4->handle_patch(id, po, &remote_aor, trail);
+    delete remote_aor; remote_aor = NULL;
+
+    if (rc == HTTP_PRECONDITION_FAILED)
+    {
+      // We've tried to do a PATCH to a remote site that doesn't have any data.
+      // We need to send a PUT.
+      TRC_DEBUG("Need to convert PATCH to PUT for %s", _id.c_str());
+      AoR* aor_for_put = new AoR(id);
+      aor_for_put->copy_aor(aor);
+      remote_s4->handle_put(id, *aor_for_put, trail);
+      delete aor_for_put; aor_for_put = NULL;
+    }
+  }
+}
+
+Store::Status S4::get_aor(const std::string& id,
+                          AoR** aor,
+                          SAS::TrailId trail)
+{
+  Store::Status rc;
+
+  *aor = _aor_store->get_aor_data(id, trail);
+
+  if (aor == NULL || *aor == NULL)
+  {
+    // Store error when getting data - return an error.
+    TRC_DEBUG("Store error when getting the AoR for %s from %s",
+               id.c_str(), _id.c_str());
+    rc = Store::Status::ERROR;
+  }
+  else if ((*aor)->bindings().empty())
+  {
+    // We successfully contacted the store, but we didn't find the AoR. This
+    // creates an empty AoR - delete this and return not found.
+    TRC_DEBUG("No AoR found for %s from %s", id.c_str(), _id.c_str());
+    rc = Store::Status::NOT_FOUND;
+    delete *aor; *aor = NULL;
+  }
+  else
+  {
+    TRC_DEBUG("Found an AoR for %s from %s", id.c_str(), _id.c_str());
+    rc = Store::Status::OK;
+  }
+
+  return rc;
+}
+
+Store::Status S4::write_aor(const std::string& id,
+                            AoR& aor,
+                            SAS::TrailId trail)
+{
+  // If the AoR has no bindings then it should be deleted. Clear up any
+  // subscriptions.
+  if (aor.bindings().empty() && !aor.subscriptions().empty())
+  {
+    TRC_DEBUG("Cleaning up AoR");
+    aor.clear(true);
+  }
+
+  int now = time(NULL);
+
+  // Send any Chronos timer requests
```

Comment:
..if we're the local site. Incidentally, I wasn't expecting this function to send Chronos timers based on the header file function comment, although I guess this work hasn't been completed yet